Allow browsing potential bug duplicates before filing a new external bug

EvaBardou opened the pull request:
Closes #681

La0 commented on server/crashmanager/serializers.py:
```diff
@@ -199,3 +199,16 @@ class Meta:
+class ExternalBugSerializer(serializers.Serializer):
+    provider = serializers.PrimaryKeyRelatedField(queryset=BugProvider.objects.all())
+    bug_id = serializers.CharField()
+    redirect_url = serializers.SerializerMethodField(required=False)
```
I would use either `url` or `details_url` as this could be used in other cases

La0 commented on server/crashmanager/views.py:
```diff
@@ -1034,6 +1034,35 @@ def partial_update(self, request, pk=None):
+        entry.bucket.bug = extBug
+        entry.bucket.save(update_fields=['bug'])
+
+        return Response(ExternalBugSerializer({'entry': entry, **serializer.validated_data}).data)
```
Please add a small comment to explain that you need to send the entry here (and that's why you need to use `update` instead of `perform_update`)

La0 commented on server/frontend/src/components/Bugs/PossibleDuplicates/List.vue:
```diff
@@ -0,0 +1,160 @@
+<template>
+  <div>
+    <button
```
Please add the input text to manage summary instead of the django form (using a bootstrap group input for display)

La0 reviewed (commented):
A few nits + the missing input, and we are good to go

La0 commented:
Deployed on https://fuzzmanager.teklia.com/crashmanager/crashes/5940/createbug/

jschwartzentruber reviewed (commented):
Is there a reason we need a new API instead of implementing `partial_update` on `Bucket`? The bucket endpoint already returns `bug.externalId` as `bug`, so I'd rather add a field on the `Bucket` serializer to take a `bug_provider` (either `classname` or `hostname`) and do a PATCH of both `bug` and `bug_provider` instead... That matches the `CrashEntry` API more closely.

The UX looks good. I did get a 500 error when clicking `Use this bug` button.

jschwartzentruber commented:
To explain further ... on `CrashEntry` we have a number of related models that we don't expose on the REST API directly, but treat them only as implementation details (`Product`, `OS`, `Platform`, `TestCase`, etc.). We expose the data for these models inline on the crashes serializer for `list`/`view`, implicitly create missing models on `create`, and update the linked `crash.testcase.quality` on `partial_update` of `testcase_quality` (on the `CrashEntry`).

We already started down this path on the `Bucket` serializer too ... as the `.bug` field on the buckets endpoint returns not `bucket.bug.pk`, but `bucket.bug.externalId`. I think it would be more consistent to do the linking operation as a `partial_update` on `Bucket` which expects `bug_provider` to exist and implicitly creates a `Bug` object.

La0 commented:
Sure, we can do that, using `BucketViewset.partial_update` and adding a `bug_provider` foreign key field on `BucketSerializer`.

It seemed easier to implement a dedicated endpoint, but you're right it looks cleaner from the API point of view.

EvaBardou commented:
> Is there a reason we need a new API instead of implementing `partial_update` on `Bucket`? The bucket endpoint already returns `bug.externalId` as `bug`, so I'd rather add a field on the `Bucket` serializer to take a `bug_provider` (either `classname` or `hostname` ... I guess using pk, since there is already a `BugProvider` serializer) and do a PATCH of both `bug` and `bug_provider` instead... That matches the `CrashEntry` API more closely.
> 
> The UX looks good. I did get a 500 error when clicking `Use this bug` button.

@jschwartzentruber I rebased the PR and updated it to use the `BucketViewSet` instead of creating a new one to assign the external Bugzilla bug, could you please review again? Thx!

P.S.: I also added a fix for another 500 error on signature list (mode: `all`) page. It occurred when the `BugProvider` linked to a `Bucket.bug` didn't have an `urlTemplate` (like on #697).

La0 commented on server/crashmanager/templates/signatures/index.html:
```diff
@@ -37,8 +37,10 @@
             <td>{{ entry.quality }}</td>
             <td>
-                {% if entry.bug %}
+                {% if entry.bug.externalType.urlTemplate %}
```
Please keep the `{% if entry.bug %}` check too, or it might fail

jschwartzentruber commented on server/crashmanager/serializers.py:
```diff
@@ -124,15 +124,20 @@ def create(self, attrs):
     # size and best_quality are annotations, so must be set manually
     size = serializers.IntegerField(write_only=True, required=False)
     best_quality = serializers.IntegerField(write_only=True, required=False)
+    provider = serializers.PrimaryKeyRelatedField(write_only=True, required=False, queryset=BugProvider.objects.all())
```
nit: I'd prefer this field to be `bug_provider`

jschwartzentruber reviewed (approved):
Looks good besides naming nit
